datafusion/physical-expr/src/sort_expr.rs

```diff
@@ -107,19 +147,87 @@ impl std::fmt::Display for PhysicalSortRequirement {
 }
 
 impl PhysicalSortRequirement {
+    /// Creates a new requirement.
+    ///
+    /// If `options` is `Some(..)`, creates an `exact` requirement,
+    /// which must match both `options` and `expr`.
+    ///
+    /// If `options` is `None`, Creates a new `expr_only` requirement,
+    /// which must match only `expr`.
+    ///
+    /// See [`PhysicalSortRequirement`] for examples.
+    pub fn new(expr: Arc<dyn PhysicalExpr>, options: Option<SortOptions>) -> Self {
+        Self { expr, options }
+    }
+
+    /// Replace the required expression for this requirement with the new one
+    pub fn with_expr(mut self, expr: Arc<dyn PhysicalExpr>) -> Self {
+        self.expr = expr;
+        self
+    }
+
+    /// Converts the `PhysicalSortRequirement` to `PhysicalSortExpr`.
+    /// If required ordering is `None` for an entry, the default
+    /// ordering `ASC, NULLS LAST` is used.
+    ///
+    /// The default is picked to be consistent with
+    /// PostgreSQL: <https://www.postgresql.org/docs/current/queries-order.html>
+    pub fn into_sort_expr(self) -> PhysicalSortExpr {
+        let Self { expr, options } = self;
+
+        let options = options.unwrap_or(SortOptions {
+            descending: false,
+            nulls_first: false,
+        });
+        PhysicalSortExpr { expr, options }
+    }
+
     /// Returns whether this requirement is equal or more specific than `other`.
     pub fn compatible(&self, other: &PhysicalSortRequirement) -> bool {
         self.expr.eq(&other.expr)
             && other.options.map_or(true, |other_opts| {
                 self.options.map_or(false, |opts| opts == other_opts)
             })
     }
-}
 
-pub fn make_sort_requirements_from_exprs(
-    ordering: &[PhysicalSortExpr],
-) -> Vec<PhysicalSortRequirement> {
-    ordering.iter().map(|e| e.clone().into()).collect()
+    /// Returns [`PhysicalSortRequirement`] that requires the exact
+    /// sort of the [`PhysicalSortExpr`]s in `ordering`
+    ///
+    /// This method is designed for
+    /// use implementing [`ExecutionPlan::required_input_ordering`].
+    pub fn from_sort_exprs<'a>(
+        ordering: impl IntoIterator<Item = &'a PhysicalSortExpr>,
+    ) -> Vec<PhysicalSortRequirement> {
+        ordering
+            .into_iter()
+            .map(PhysicalSortRequirement::from)
+            .collect()
+    }
+
+    /// Converts an iterator of [`PhysicalSortRequirement`] into a Vec
+    /// of [`PhysicalSortExpr`]s.
+    ///
+    /// This function converts `PhysicalSortRequirement` to `PhysicalSortExpr`
+    /// for each entry in the input. If required ordering is None for an entry
+    /// default ordering `ASC, NULLS LAST` if given (see [`into_sort_expr`])
+    pub fn to_sort_exprs(
+        requirements: impl IntoIterator<Item = PhysicalSortRequirement>,
+    ) -> Vec<PhysicalSortExpr> {
```

Comment:
I think we can use     
```rust
pub fn to_sort_exprs<'a>(
        requirements: impl IntoIterator<Item = &'a PhysicalSortRequirement>,
    ) -> Vec<PhysicalSortExpr> 
```
as signature for `to_sort_exprs` as in `from_sort_exprs`. This would help remove a couple clones in the code.

Reply:
My rationale for taking `PhysicalSortRequirement` rather than `&PhysicalSortRequirement` was actually to avoid a clone in the cases where the caller already had an owned `PhysicalSortRequirement`. Basically the clone() is moved from inside this function to the callsite

Specifically in `datafusion/core/src/physical_optimizer/sort_enforcement.rs`

```rust
                    // Make sure we preserve the ordering requirements:
                    update_child_to_remove_unnecessary_sort(child, sort_onwards, &plan)?;
                    let sort_expr =
                        PhysicalSortRequirement::to_sort_exprs(required_ordering);
                    add_sort_above(child, sort_expr)?;
                    if is_sort(child) {
``` 
This signature can avoid a clone. 


If the signature looks like 

```rust
pub fn to_sort_exprs<'a>(
        requirements: impl IntoIterator<Item = &'a PhysicalSortRequirement>,
    ) -> Vec<PhysicalSortExpr> 
```

I think it would  require always `clone()`ing the expr and options internally to create an owned `PhysicalSortExpr` in the output

Of course, we are talking about cloning an `Arc` and some `SortOption`s which presumably isn't all that expensive , so maybe it doesn't matter 🤔

Reply:
I have a feeling there may be a way to do this with some `Borrow` trick to implement one function who deals with owned and referenced cases without duplicating code 🤔

Reply:
@alamb, it seems like using
```rust
pub fn to_sort_exprs<T: Borrow<PhysicalSortRequirement>>(
    requirements: impl IntoIterator<Item = T>,
) -> Vec<PhysicalSortExpr> {
    requirements
        .into_iter()
        .map(|e| PhysicalSortExpr::from(e.borrow()))
        .collect()
}

pub fn from_sort_exprs<T: Borrow<PhysicalSortExpr>>(
    ordering: impl IntoIterator<Item = T>,
) -> Vec<PhysicalSortRequirement> {
    ordering
        .into_iter()
        .map(|e| PhysicalSortRequirement::from(e.borrow()))
        .collect()
}
```
lets us get rid of a bunch of `clone`s and results in simpler call site code. AFAIK, this is a zero-cost mechanism. If you concur, we can use this pattern.

I created [#11](https://github.com/alamb/arrow-datafusion/pull/11) on your repo so you can easily check it out and incorporate.

Reply:
I guess my point was that I think this call:

```rust
PhysicalSortRequirement::from(e.borrow()
```

Invokes this `From` impl:
```
impl From<&PhysicalSortRequirement> for PhysicalSortExpr {
    fn from(value: &PhysicalSortRequirement) -> Self {
        value.clone().into_sort_expr()
    }
}
```

Which has a `clone` of the value

I have removed the `From<&PhysicalSortRequirement>` traits in 12cecdb4e because it is obscuring the locations of the cloning.

Reply:
I see. I thought you wanted to make `to_sort_exprs` "smart" in the sense that it can take work with an owned value or a reference, using the owned or reference `From` impl depending on the argument -- making a clone only when necessary.

Anyway I got curious and tried to verify whether the `borrow` mechanism was zero-cost for this purpose; it turns out it is not. It always calls the reference path. I also found how to do this -- here is an [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=21c9901d73b494c02be88157306f7bf0) for future reference.

For posterity, had we wanted to go this route, the smart `to_sort_exprs` would be:
```rust
pub fn to_sort_exprs<T>(
    requirements: impl IntoIterator<Item = T>,
) -> Vec<PhysicalSortExpr>
where
    PhysicalSortExpr: From<T>,
{
    requirements
        .into_iter()
        .map(|e| PhysicalSortExpr::from(e))
        .collect()
}
```
which would route to the right `PhysicalSortExpr::from` depending on the argument.

Reply:
Very cool -- I did not know that. If you think that is a worthwhile change, I would be happy to make a follow on PR

Reply:
I think it'd be a good change, I can already see some use cases for it in some upcoming code (e.g. streaming group bys).

I updated the mini-PR I created for experimentation so you can get the changes.

Reply:
Thanks -- I tried to pull the commits into https://github.com/apache/arrow-datafusion/pull/5918 but they still seem to be trying to use this impl:

```
impl From<&PhysicalSortRequirement> for PhysicalSortExpr {
```

I may have messed up the merge somehow

Let's keep collaborating on https://github.com/apache/arrow-datafusion/pull/5918